ci/Dockerfile-envoy

```diff
@@ -1,88 +1,76 @@
-
+FROM ${BUILD_OS}:${BUILD_TAG} AS envoy-base
 ENV DEBIAN_FRONTEND=noninteractive
-
+RUN mkdir -p /etc/envoy
+RUN adduser --group --system envoy
 RUN apt-get -qq update && apt-get -qq upgrade -y \
     && apt-get -qq install --no-install-recommends -y ca-certificates \
     && apt-get -qq autoremove -y && apt-get clean \
     && rm -rf /tmp/* /var/tmp/* \
     && rm -rf /var/lib/apt/lists/*
+EXPOSE 10000
+ENTRYPOINT ["/docker-entrypoint.sh"]
+CMD ["envoy", "-c", "/etc/envoy/envoy.yaml"]
 
-RUN mkdir -p /etc/envoy
 
-COPY --from=binary /usr/local/bin/envoy* /usr/local/bin/
-COPY --from=binary /usr/local/bin/su-exec /usr/local/bin/
+# STAGE: envoy
+FROM envoy-base AS envoy
 COPY --from=binary /etc/envoy/envoy.yaml /etc/envoy/envoy.yaml
 COPY --from=binary /docker-entrypoint.sh /
-
-RUN adduser --group --system envoy
-
-EXPOSE 10000
-
-ENTRYPOINT ["/docker-entrypoint.sh"]
-CMD ["envoy", "-c", "/etc/envoy/envoy.yaml"]
+COPY --from=binary /usr/local/bin/su-exec /usr/local/bin/
+COPY --from=binary /usr/local/bin/envoy* /usr/local/bin/
 
 
 # STAGE: envoy-distroless
 # gcr.io/distroless/base-nossl-debian11:nonroot
 FROM gcr.io/distroless/base-nossl-debian11:nonroot@sha256:8c880faa0cb8c3fe60cae75de2cf4f6e7b53bd977a351a7e5ab510394e509f24 AS envoy-distroless
-
-COPY --from=binary /usr/local/bin/envoy* /usr/local/bin/
-COPY --from=binary /etc/envoy/envoy.yaml /etc/envoy/envoy.yaml
-
 EXPOSE 10000
-
 ENTRYPOINT ["/usr/local/bin/envoy"]
 CMD ["-c", "/etc/envoy/envoy.yaml"]
+COPY --from=binary /etc/envoy/envoy.yaml /etc/envoy/envoy.yaml
+COPY --from=binary /usr/local/bin/envoy* /usr/local/bin/
 
 
 # STAGE: envoy-google-vrp-base
 FROM ${ENVOY_VRP_BASE_IMAGE} AS envoy-google-vrp-base
-
-ENV DEBIAN_FRONTEND=noninteractive
-
+EXPOSE 10000
+EXPOSE 10001
+CMD ["supervisord", "-c", "/etc/supervisor.conf"]
+ENTRYPOINT []
+ADD configs/google-vrp/envoy-edge.yaml /etc/envoy/envoy-edge.yaml
+ADD configs/google-vrp/envoy-origin.yaml /etc/envoy/envoy-origin.yaml
+ADD configs/google-vrp/launch_envoy.sh /usr/local/bin/launch_envoy.sh
+ADD test/config/integration/certs/serverkey.pem /etc/envoy/certs/serverkey.pem
+ADD test/config/integration/certs/servercert.pem /etc/envoy/certs/servercert.pem
 RUN apt-get -qq update \
     && apt-get -qq upgrade -y \
     && apt-get -qq install -y libc++1 supervisor gdb strace tshark \
     && apt-get autoremove -y \
     && apt-get clean \
     && rm -rf /tmp/* /var/tmp/* \
     && rm -rf /var/lib/apt/lists/*
-
-ADD configs/google-vrp/envoy-edge.yaml /etc/envoy/envoy-edge.yaml
-ADD configs/google-vrp/envoy-origin.yaml /etc/envoy/envoy-origin.yaml
-ADD configs/google-vrp/launch_envoy.sh /usr/local/bin/launch_envoy.sh
 ADD configs/google-vrp/supervisor.conf /etc/supervisor.conf
-ADD test/config/integration/certs/serverkey.pem /etc/envoy/certs/serverkey.pem
-ADD test/config/integration/certs/servercert.pem /etc/envoy/certs/servercert.pem
 RUN chmod 777 /var/log/supervisor
 RUN chmod a+r /etc/supervisor.conf /etc/envoy/* /etc/envoy/certs/*
```

Comment:
not sure what this supervisor is, shouldn't its config setup be done before calling it?

Reply:
process handler - like a mini systemd

> shouldn't its config setup be done before calling it?

these are instructinos/layers in a `Dockerfile` so its fine to add a layer that says when this container runs do this, and another layer after that says - this is what the container will need

Reply:
ok, I thought it was RUN for a moment.

Reply:
this one ive tested a lot more, so pretty confident this what we need - if we cache the `/var/lib/docker` on the docker_ci stage we can make it run almost entirely off cache and just run the "add envoy" layers